posthog/special_migrations/definition.py

```diff
@@ -1,34 +1,81 @@
-from typing import List, Optional, Tuple
+from datetime import datetime
+from typing import Callable, List, Optional, Tuple
 
-from posthog.models.special_migration import SpecialMigration
+from posthog.constants import AnalyticsDBMS
 from posthog.version_requirement import ServiceVersionRequirement
 
 
+# used to prevent circular imports
+class SpecialMigrationType:
+    id: int
+    name: str
+    description: str
+    progress: int
+    status: int
+    current_operation_index: int
+    current_query_id: str
+    celery_task_id: str
+    started_at: str
+    finished_at: datetime
+    last_error: str
+    posthog_min_version: str
+    posthog_max_version: str
+
+
 class SpecialMigrationOperation:
-    def __init__(self, sql="", database="clickhouse", timeout_seconds=60, rollback=None, resumable=False):
+    def __init__(
+        self,
+        sql="",
+        database: AnalyticsDBMS = AnalyticsDBMS.CLICKHOUSE,
+        timeout_seconds: int = 60,
+        rollback="",
+        resumable=False,
+    ):
         self.sql = sql
         self.database = database
+
+        # currently CH only
         self.timeout_seconds = timeout_seconds
+
+        # if the operation is dynamic and knows how to restart correctly after a crash
+        # Example:
+        #   - Not resumable: `INSERT INTO table1 (col1) SELECT col1 FROM table2`
+        #   - Resumable: `INSERT INTO table2 (foo, timestamp) SELECT foo, timestamp FROM table1 WHERE timestamp > (SELECT max(timestamp) FROM table2)`
         self.resumable = resumable
+
+        # This should not be a long operation as it will be executed synchronously!
         self.rollback = rollback
 
 
 class SpecialMigrationDefinition:
+
+    # the migration cannot be run before this version
     posthog_min_version = "0.0.0"
+
+    # the migration _must_ be run before this version
     posthog_max_version = "10000.0.0"
 
+    # use this to add information about why this migration is needed to self-hosted users
+    description = ""
+
+    # list of versions accepted for the services the migration relies on e.g. ClickHouse, Postgres
     service_version_requirements: List[ServiceVersionRequirement] = []
+
+    # list of operations the migration will perform _in order_
     operations: List[SpecialMigrationOperation] = []
-    dependencies: List[str]
 
+    # name of special migration this migration depends on
+    depends_on: Optional[str] = None
```

Comment:
This is super solid. My only questions are:

- What if the special migration depends on a normal migration?
- What if a normal migration depends on a special migration?

Reply:
No such things supported for now. The current way (which I think makes sense) is that we specify PostHog version ranges. Thus, the normal migrations it depends on are those required for the min version it specifies.

And a normal migration wouldn't depend on a special migration, it would depend on a PostHog version as well, which is the max version specified by it.

We can definitely change this in the future if we need more granularity